File: backend/append_geojson.js
require('dotenv').config();
const fs = require('fs');
const { createClient } = require('@supabase/supabase-js');

const supabase = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_SERVICE_ROLE_KEY);

async function run() {
    try {
        console.log('Reading berbera_buildings.geojson...');
        const raw = fs.readFileSync('berbera_buildings.geojson', 'utf8');
        const geojson = JSON.parse(raw);
        const features = geojson.features || [];

        console.log(`Loaded ${features.length} buildings.`);

        const { data: maxRow } = await supabase
            .from('buildings')
            .select('number')
            .order('number', { ascending: false })
            .limit(1)
            .single();

        let number = (maxRow?.number || 0) + 1;
        console.log(`Starting insertion from number ${number}...`);

        const rows = features.map(feat => {
            let coords = feat.geometry.coordinates;
            if (feat.geometry.type === 'MultiPolygon') coords = coords[0][0][0];
            else if (feat.geometry.type === 'Polygon') coords = coords[0][0];
            else if (feat.geometry.type === 'Point') coords = coords;

            return {
                number: number++,
                longitude: coords[0],
                latitude: coords[1]
            };
        });

        console.log(`Inserting ${rows.length} rows...`);
        const CHUNK = 1000;
        let inserted = 0;
        for (let i = 0; i < rows.length; i += CHUNK) {
            const chunk = rows.slice(i, i + CHUNK);
            const { error } = await supabase.from('buildings').upsert(chunk, { onConflict: 'number' });
            if (error) {
                console.error('Error inserting chunk:', error);
                throw error;
            }
            inserted += chunk.length;
            if (inserted % 2000 === 0) console.log(`Inserted ${inserted} buildings...`);
        }

        console.log('✅ Successfully appended all Berbera buildings from geojson!');
    } catch (err) {
        console.error('Script failed:', err);
    }
}

run();

File: backend/append_geojson.py
import json
import os
from dotenv import load_dotenv
from supabase import create_client
from shapely.geometry import shape

load_dotenv()

supabase_url = os.environ.get("SUPABASE_URL")
supabase_key = os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
supabase = create_client(supabase_url, supabase_key)

file_path = "berbera_buildings.geojson"
print(f"Reading {file_path}...")
with open(file_path, "r", encoding="utf-8") as f:
    geojson = json.load(f)
    
features = geojson.get("features", [])
print(f"Loaded {len(features)} buildings from {file_path}.")

# Get max number safely so we don't wipe existing DB
response = supabase.table("buildings").select("number").order("number", desc=True).limit(1).execute()
if response.data:
    current_max = response.data[0]["number"]
else:
    current_max = 0

number = current_max + 1
print(f"Found current max number {current_max}. Starting insertion from number {number}...")

rows = []
for feat in features:
    geom = shape(feat["geometry"])
    center = geom.centroid # Extract exact lat/lng point from Google Open Buildings polygon
    
    rows.append({
        "number": number,
        "longitude": round(center.x, 8),
        "latitude": round(center.y, 8)
    })
    number += 1

print(f"Inserting {len(rows)} buildings...")
CHUNK = 1000
for i in range(0, len(rows), CHUNK):
    chunk = rows[i:i+CHUNK]
    # use upsert to cleanly append and avoid errors
    res = supabase.table("buildings").upsert(chunk, on_conflict="number").execute()
    print(f"Inserted up to {i + len(chunk)} / {len(rows)}...")

print("✅ Successfully appended all Berbera buildings!")
